src/Collection.php

```diff
@@ -658,6 +685,10 @@ public function find($filter = [], array $options = [])
      */
     public function findOne($filter = [], array $options = [])
     {
+        if (! array_key_exists('codec', $options)) {
+            $options['codec'] = $this->codec;
+        }
```

Comment:
I'll take a look at how we can support this. For operations that return documents (either in a Cursor or a single document) this may be feasible, but for other operations like `replace` and `insert` that encode incoming values it's a little more complicated. Incoming values are passed in the constructor to the parent operation, so the middleware can't just wrap an existing operation and modify its return value. Furthermore, the type map option needs to be updated accordingly, which means that the middleware would have to be the one creating the original operation in order to encode incoming values and update the type map accordingly.

Reply:
Oh, I wasn't actually suggesting middleware to handle documents on the way in/out. It was more about wrapping operations with common behaviors, such as inheriting an option or doing transaction checks. So much of the code in the helpers is repeated option handling and it's mostly pre-execution since we often return the result of `execute()` directly.

Reply:
Ah, I see, so I misunderstood. I believe the key here would be to have distinct classes for options as other drivers already have them. This would allow us to use types to cut down on such checks and also add custom logic to each options class while keeping common logic reusable using traits. I can draft such a change so we know the impact and can decide when and how to make it.

Reply:
SGTM. I've long wondered if there was a use case for using something like Symfony's [OptionsResolver component](https://symfony.com/doc/current/components/options_resolver.html) but was worried the overhead might not be worth it.

I definitely think there's some room for improvement vs. what we have now, though. Would be helpful to revisit after get benchmarks in place and can do some profiling.

Reply:
Extracting options handling to separate classes is being tracked in PHPLIB-1210.